webhooks/cluster_test.go

```diff
@@ -410,3 +432,643 @@ func TestClusterTopologyValidationWithClient(t *testing.T) {
 		})
 	}
 }
+
+// TestClusterTopologyValidationForTopologyClassChange cases where cluster.spec.topology.class is altered.
+func TestClusterTopologyValidationForTopologyClassChange(t *testing.T) {
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+	g := NewWithT(t)
+
+	ref := &corev1.ObjectReference{
+		APIVersion: "group.test.io/foo",
+		Kind:       "barTemplate",
+		Name:       "baz",
+		Namespace:  "default",
+	}
+	compatibleNameChangeRef := &corev1.ObjectReference{
+		APIVersion: "group.test.io/foo",
+		Kind:       "barTemplate",
+		Name:       "differentbaz",
+		Namespace:  "default",
+	}
+	compatibleAPIVersionChangeRef := &corev1.ObjectReference{
+		APIVersion: "group.test.io/foo2",
+		Kind:       "barTemplate",
+		Name:       "differentbaz",
+		Namespace:  "default",
+	}
+	incompatibleKindRef := &corev1.ObjectReference{
+		APIVersion: "group.test.io/foo",
+		Kind:       "another-barTemplate",
+		Name:       "another-baz",
+		Namespace:  "default",
+	}
+	incompatibleAPIGroupRef := &corev1.ObjectReference{
+		APIVersion: "group.nottest.io/foo",
+		Kind:       "barTemplate",
+		Name:       "another-baz",
+		Namespace:  "default",
+	}
+
+	tests := []struct {
+		name        string
+		cluster     *clusterv1.Cluster
+		firstClass  *clusterv1.ClusterClass
+		secondClass *clusterv1.ClusterClass
+		wantErr     bool
+	}{
+		// InfrastructureCluster changes.
+		{
+			name: "Accept cluster.topology.class change with a compatible infrastructureCluster Kind ref change",
+			cluster: builder.Cluster(metav1.NamespaceDefault, "cluster1").
```

Comment:
If I saw it right the cluster is the same for all test cases. In this case I would suggest dropping the cluster from the table. I think that's fine as the assumption that the second class is called "class2" and we're migration from "class1" to "class2" is already baked in into the for loop. So there is no harm in doing the same for the Cluster and it helps to focus on the diff between the ClusterClasses when reading the table tests

Reply:
Good idea!